In [1]:
import pandas as pd
import openpyxl
import sys
sys.path.append('../')
from src import sp_eda as se
pd.set_option('display.max_columns', None)

### Insertamos los datos del csv


In [2]:
df1=pd.read_csv('../data/bank-additional.csv')


In [3]:
se.eda_prelim(df1)

,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
23023,23120,56.0,retired,MARRIED,professional.course,0.0,0.0,0.0,cellular,139,2,999,0,NONEXISTENT,1.4,"93,444","-36,1",NaN,"5228,1",no,1-mayo-2016,39.638,-66.979,ba118301-a3de-46d7-92c4-271d1d2ee024
32080,32224,36.0,blue-collar,MARRIED,basic.9y,0.0,0.0,1.0,cellular,49,1,999,0,NONEXISTENT,-1.8,"92,893","-46,2","1,313","5099,1",no,23-octubre-2015,35.972,-105.105,832d4428-6026-4566-ade2-fab98679e6c9
29393,29537,NaN,blue-collar,MARRIED,basic.9y,0.0,0.0,0.0,cellular,98,3,999,0,NONEXISTENT,-1.8,"93,075","-47,1",NaN,"5099,1",no,24-enero-2019,45.751,-121.548,12f0efd9-f39b-450b-bca4-974c006751ee
33861,34005,NaN,blue-collar,MARRIED,basic.6y,0.0,0.0,1.0,cellular,281,1,999,0,NONEXISTENT,-1.8,"92,893","-46,2","1,281","5099,1",no,16-julio-2019,40.632,-76.601,906b6f2a-4f8b-4870-8372-ff25043a8ac4
28413,28557,53.0,retired,MARRIED,high.school,0.0,0.0,0.0,cellular,349,1,999,0,NONEXISTENT,-1.8,"93,075","-47,1","1,415","5099,1",no,7-junio-2016,47.742,-70.551,f024d7c3-cea6-4648-90a3-960ec722d45a


----------
DIMENSIONES
El conjunto de datos presenta 43000 filas y 24 columnas
----------
INFORMACION DE COLUMNAS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.pr

None

----------
DUPLICADOS
Tenemos 0 duplicados
----------
ELEMENTOS NULOS


Unnamed: 0         0.000000
age               11.906977
job                0.802326
marital            0.197674
education          4.202326
default           20.886047
housing            2.386047
loan               2.386047
contact            0.000000
duration           0.000000
campaign           0.000000
pdays              0.000000
previous           0.000000
poutcome           0.000000
emp.var.rate       0.000000
cons.price.idx     1.095349
cons.conf.idx      0.000000
euribor3m         21.525581
nr.employed        0.000000
y                  0.000000
date               0.576744
latitude           0.000000
longitude          0.000000
id_                0.000000
dtype: float64

----------
FRECUENCIA CATEGORÍAS
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
-----------
marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64
-----------
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
-----------
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
-----------
poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64
-----------
cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201 

### Insertamos los datos del xlxs, incorporando las distintas hojas en el mismo dataframe


In [4]:
hojas=pd.read_excel('../data/customer-details.xlsx',sheet_name=None)

In [5]:
df_details_completo= pd.concat(hojas.values(),ignore_index=True)


In [6]:
se.eda_prelim(df_details_completo)

,Unnamed: 0,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
11384,11384,79031,0,2,2012-11-06,16,27fc8fef-9f3e-443f-adc8-3a36413e94df
26215,6100,73417,2,2,2013-12-20,32,35aa2907-61ed-4044-be4e-dca931c937a4
42159,13079,74132,2,2,2014-11-05,28,32401cb2-a49d-4a2c-a9ba-a853181df198
36090,7010,74716,0,0,2014-02-25,21,8d57a062-00d1-4454-a13c-89d95103e115
15929,15929,103389,1,2,2012-05-20,20,e8992245-7cd3-4b90-9771-86b228d4c3ee


----------
DIMENSIONES
El conjunto de datos presenta 43170 filas y 7 columnas
----------
INFORMACION DE COLUMNAS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Unnamed: 0         43170 non-null  int64         
 1   Income             43170 non-null  int64         
 2   Kidhome            43170 non-null  int64         
 3   Teenhome           43170 non-null  int64         
 4   Dt_Customer        43170 non-null  datetime64[ns]
 5   NumWebVisitsMonth  43170 non-null  int64         
 6   ID                 43170 non-null  object        
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 2.3+ MB


None

----------
DUPLICADOS
Tenemos 0 duplicados
----------
ELEMENTOS NULOS


Unnamed: 0           0.0
Income               0.0
Kidhome              0.0
Teenhome             0.0
Dt_Customer          0.0
NumWebVisitsMonth    0.0
ID                   0.0
dtype: float64

----------
FRECUENCIA CATEGORÍAS
ID
089b39d8-e4d0-461b-87d4-814d71e0e079    1
38708e69-8996-4d51-956a-d422fd78872d    1
7218c062-4571-4fd9-8bad-c6a09ad1eb5d    1
853e4074-829e-47e4-9287-06b4cb0f814f    1
a185385b-b614-423b-9932-a132ed02eed1    1
                                       ..
8e1e500d-de2a-4e14-b027-20983ebaeeb9    1
bdee816d-7f9d-442a-99e1-cf259021f8f5    1
8d3ae2ee-5363-4508-ac65-cbecc78a08bb    1
4601e86a-18c4-473c-8467-fb16e838be7d    1
9727bc06-c11a-461a-a5bb-3d210467cc2a    1
Name: count, Length: 43170, dtype: int64
-----------


###  Unimos los dos dataframe 
Los vamos a unir a traves de la columna id_. Para ello nos damos cuenta que en df_details_completo la columna ID representa el identificador del cliente al igual que en la columna id_ del df1. Cambiamos el nombre para que coincida y podamos unir los dos dataframes por esa misma columna.Eliminamos las columnas Unnamed porvenientes de los dos df originales.

In [7]:
df_details_completo.rename(columns={'ID':'id_'}, inplace= True)

df=pd.merge(df1,df_details_completo,'outer', 'id_')
df=df.drop(columns=['Unnamed: 0_x','Unnamed: 0_y'])

In [8]:
se.eda_prelim(df)

,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth
3267,51.0,admin.,MARRIED,high.school,0.0,1.0,0.0,cellular,326.0,2.0,999.0,0.0,NONEXISTENT,-0.1,"93,2",-42,"4,191","5195,8",no,7-febrero-2019,40.551,-116.709,132a0ee8-df51-43ab-8daf-2309e40234f1,157830,0,2,2013-11-09,12
534,33.0,entrepreneur,MARRIED,high.school,0.0,1.0,0.0,cellular,902.0,2.0,999.0,0.0,NONEXISTENT,-0.1,"93,2",-42,"4,12","5195,8",no,17-abril-2019,37.721,-86.967,03259cd8-f959-4829-b023-8c76dbe67185,111815,1,0,2013-03-11,10
24893,39.0,blue-collar,SINGLE,basic.9y,0.0,1.0,0.0,cellular,340.0,1.0,999.0,0.0,NONEXISTENT,-1.8,"93,075","-47,1","1,483","5099,1",no,25-agosto-2018,42.487,-99.740,93838f0b-b6b7-41c9-8470-69986d437a62,115524,2,1,2013-01-29,27
18338,32.0,admin.,DIVORCED,high.school,0.0,0.0,0.0,cellular,115.0,1.0,999.0,1.0,FAILURE,-0.1,"93,2",-42,"4,153","5195,8",no,3-mayo-2018,36.876,-70.430,6cc0213a-e2f0-40e4-b6ef-6ed7a15f4d37,30408,2,1,2013-11-02,2
16289,46.0,admin.,DIVORCED,high.school,0.0,1.0,1.0,telephone,233.0,4.0,999.0,0.0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,4-marzo-2019,29.284,-83.564,6099da9d-24f4-497a-a59a-d7ea2484e3ea,133240,1,0,2012-07-22,8


----------
DIMENSIONES
El conjunto de datos presenta 43170 filas y 28 columnas
----------
INFORMACION DE COLUMNAS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   age                37880 non-null  float64       
 1   job                42655 non-null  object        
 2   marital            42915 non-null  object        
 3   education          41193 non-null  object        
 4   default            34019 non-null  float64       
 5   housing            41974 non-null  float64       
 6   loan               41974 non-null  float64       
 7   contact            43000 non-null  object        
 8   duration           43000 non-null  float64       
 9   campaign           43000 non-null  float64       
 10  pdays              43000 non-null  float64       
 11  previous           43000 non-null  float64       
 12  po

None

----------
DUPLICADOS
Tenemos 0 duplicados
----------
ELEMENTOS NULOS


age                  12.253880
job                   1.192958
marital               0.590688
education             4.579569
default              21.197591
housing               2.770442
loan                  2.770442
contact               0.393792
duration              0.393792
campaign              0.393792
pdays                 0.393792
previous              0.393792
poutcome              0.393792
emp.var.rate          0.393792
cons.price.idx        1.484827
cons.conf.idx         0.393792
euribor3m            21.834607
nr.employed           0.393792
y                     0.393792
date                  0.968265
latitude              0.393792
longitude             0.393792
id_                   0.000000
Income                0.000000
Kidhome               0.000000
Teenhome              0.000000
Dt_Customer           0.000000
NumWebVisitsMonth     0.000000
dtype: float64

----------
FRECUENCIA CATEGORÍAS
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
-----------
marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64
-----------
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
-----------
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
-----------
poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64
-----------
cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201 

Se observan porcentajes de nulos con el mismo valor en muchas columnas. Se cae en la cuenta de que hemos juntado archivos con distinto numero de registos 43000 vs 43170. los 170 de diferencia sobre 43170 representan el 0,39% mostrado en los elementos nulos. Si esos 170 registros de diferencia no tienen datos de comunicación en la campaña, quizas no tenga sentido incluirlos en el análisis. Se propone una unión de dataframes inner o left 

In [9]:
df=pd.merge(df1,df_details_completo,'inner', 'id_')
df=df.drop(columns=['Unnamed: 0_x','Unnamed: 0_y'])
se.eda_prelim(df)

,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth
29848,29.0,technician,SINGLE,university.degree,0.0,0.0,0.0,cellular,95,1,999,0,NONEXISTENT,-1.8,"93,075","-47,1","1,405","5099,1",no,5-junio-2016,45.518,-99.662,9aba67ee-8aa0-4c01-b4ba-3eed62f8a9fd,16533,2,2,2014-03-04,4
25051,38.0,admin.,MARRIED,high.school,NaN,1.0,1.0,cellular,65,2,999,0,NONEXISTENT,-0.1,"93,2",-42,NaN,"5195,8",no,3-noviembre-2018,44.716,-112.833,f3d75d74-8c51-4271-b3dc-479dfe9bd547,89046,2,0,2013-04-12,31
5518,35.0,services,MARRIED,high.school,0.0,0.0,1.0,telephone,116,1,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,26-noviembre-2017,31.597,-101.257,97bea60b-4a20-46dc-99bc-04e2a1b51574,37233,1,1,2012-05-10,14
24840,48.0,admin.,MARRIED,university.degree,0.0,1.0,1.0,cellular,122,1,999,1,FAILURE,-0.1,"93,2",-42,"4,153","5195,8",no,31-diciembre-2015,36.840,-103.060,cca3ebe8-bd84-40a5-a0c0-536eb925f195,113897,2,0,2013-05-01,14
39062,34.0,admin.,SINGLE,university.degree,0.0,1.0,0.0,cellular,265,2,999,0,NONEXISTENT,-1.8,"93,369","-34,8",NaN,"5008,7",no,26-octubre-2019,24.523,-84.266,319dd8ba-367f-4ac0-a340-3c8665c91cc8,116674,1,2,2014-10-02,6


----------
DIMENSIONES
El conjunto de datos presenta 43000 filas y 28 columnas
----------
INFORMACION DE COLUMNAS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   age                37880 non-null  float64       
 1   job                42655 non-null  object        
 2   marital            42915 non-null  object        
 3   education          41193 non-null  object        
 4   default            34019 non-null  float64       
 5   housing            41974 non-null  float64       
 6   loan               41974 non-null  float64       
 7   contact            43000 non-null  object        
 8   duration           43000 non-null  int64         
 9   campaign           43000 non-null  int64         
 10  pdays              43000 non-null  int64         
 11  previous           43000 non-null  int64         
 12  po

None

----------
DUPLICADOS
Tenemos 0 duplicados
----------
ELEMENTOS NULOS


age                  11.906977
job                   0.802326
marital               0.197674
education             4.202326
default              20.886047
housing               2.386047
loan                  2.386047
contact               0.000000
duration              0.000000
campaign              0.000000
pdays                 0.000000
previous              0.000000
poutcome              0.000000
emp.var.rate          0.000000
cons.price.idx        1.095349
cons.conf.idx         0.000000
euribor3m            21.525581
nr.employed           0.000000
y                     0.000000
date                  0.576744
latitude              0.000000
longitude             0.000000
id_                   0.000000
Income                0.000000
Kidhome               0.000000
Teenhome              0.000000
Dt_Customer           0.000000
NumWebVisitsMonth     0.000000
dtype: float64

----------
FRECUENCIA CATEGORÍAS
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
-----------
marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64
-----------
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
-----------
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
-----------
poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64
-----------
cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201 

Guardamos nuestro nuevo archivo con ambos archivos juntos y pasamos a la limpieza

In [10]:
df.to_csv('../data/b_additional_details.csv', index=False)